File: feature-promo/src/main/java/com/levnovikov/feature_promo/promo_list/di/PromoListComponent.java
package com.levnovikov.feature_promo.promo_list.di;

import com.levnovikov.feature_promo.promo_list.PromoListBuilder;
import com.levnovikov.feature_promo.promo_list.PromoListPresenter;
import com.levnovikov.feature_promo.promo_list.PromoListRouter;
import com.levnovikov.feature_promo.promo_list.PromoListView;
import com.levnovikov.feature_promo.promo_list.dependency.PromoListDependency;

import dagger.Component;
import dagger.Module;
import dagger.Provides;

/**
 * Author: lev.novikov
 * Date: 4/1/18.
 */

@PromoListScope
@Component(dependencies = PromoListDependency.class, modules = PromoListComponent.PromoListModule.class)
public interface PromoListComponent {

    void inject(PromoListView view);

    PromoListRouter router();

    void inject(PromoListBuilder promoListBuilder);

    @Module
    class PromoListModule{

        private final PromoListView view;

        public PromoListModule(PromoListView view) {
            this.view = view;
        }

        @Provides
        @PromoListScope
        PromoListView provideView() {
            return view;
        }

        @Provides
        @PromoListScope
        PromoListPresenter providePresenter() {
            return view;
        }
    }
}

File: feature-promo/src/main/java/com/levnovikov/feature_promo/promo_list/di/PromoListComponent.kt
package com.levnovikov.feature_promo.promo_list.di

import com.levnovikov.feature_promo.promo_list.PromoListNodeHolder
import com.levnovikov.feature_promo.promo_list.PromoListPresenter
import com.levnovikov.feature_promo.promo_list.PromoListRouter
import com.levnovikov.feature_promo.promo_list.PromoListView
import com.levnovikov.feature_promo.promo_list.dependency.PromoListDependency

import dagger.Component
import dagger.Module
import dagger.Provides

/**
 * Author: lev.novikov
 * Date: 4/1/18.
 */

@PromoListScope
@Component(dependencies = [(PromoListDependency::class)], modules = [(PromoListComponent.PromoListModule::class)])
interface PromoListComponent {

    fun inject(view: PromoListView)

    fun router(): PromoListRouter

    fun inject(promoListBuilder: PromoListNodeHolder)

    @Module
    class PromoListModule(private val view: PromoListView) {

        @Provides
        @PromoListScope
        internal fun provideView(): PromoListView {
            return view
        }

        @Provides
        @PromoListScope
        internal fun providePresenter(): PromoListPresenter {
            return view
        }
    }
}
